### ISSUE TITLE
recontructed images are differenct from original image in self-learning for BeitForMaskedImageModeling

### ISSUE THREAD

**Issue by vince2003** [microsoft/unilm#401]
Dear Authors,
I want to reconstruct the partially masked image to the original image. However, the final results are different from the original image.
I use: +) BeitForMaskedImageModeling for Encoder (from https://huggingface.co/transformers/master/model_doc/beit.html)
          +) Decoder from dall_e (from https://github.com/openai/DALL-E)
1) Can you tell me why the reconstructed image and original image are the big differences?
2) Can you upload the final checkpoint of the decoder in self-learning for reconstructing masked images
[beit_dall_simple.md](https://github.com/microsoft/unilm/files/7020901/beit_dall_simple.md)
?

**Comment by vince2003** [comment]
This is my code

> Dear Authors,
> I want to reconstruct the partially masked image to the original image. However, the final results are different from the original image.
> I use: +) BeitForMaskedImageModeling for Encoder (from https://huggingface.co/transformers/master/model_doc/beit.html)
> +) Decoder from dall_e (from https://github.com/openai/DALL-E)
> 
> 1. Can you tell me why the reconstructed image and original image are the big differences?
> 2. Can you upload the final checkpoint of the decoder in self-learning for reconstructing masked images
>    [beit_dall_simple.md](https://github.com/microsoft/unilm/files/7020901/beit_dall_simple.md)
>    ?

**Comment by NielsRogge** [comment]
Hi, I implemented `BeitForMaskedImageModeling`. It returns the exact same logits as the original implementation for the same `pixel_values` and `bool_masked_pos`. However, when testing it out on an image, none of the visual tokens it predicts correspond to the ground-truth visual tokens from DALL-E's tokenizer. 

@donglixp would be great if you can take a look at my notebook: https://colab.research.google.com/drive/1Mjt-3jHw9HYMXECmSdDlbiG59ZAw-Z0T?usp=sharing

**Comment by donglixp** [comment]
@addf400 could look at the above notebook for double-checking.

**Comment by vvvm23** [comment]
Is the OpenAI public decoder (`https://cdn.openai.com/dall-e/decoder.pkl`) perhaps slightly different to the one used in this work? I am having the same issue where the reconstructed outputs from `BeitForMaskedImageModeling` are much lower quality than the input, even when there is no masking applied. However, some similarity is there. I'd be interested to know if there has been any progress with this ~

**Comment by donglixp** [comment]
> Is the OpenAI public decoder (`https://cdn.openai.com/dall-e/decoder.pkl`) perhaps slightly different to the one used in this work? I am having the same issue where the reconstructed outputs from `BeitForMaskedImageModeling` are much lower quality than the input, even when there is no masking applied. However, some similarity is there. I'd be interested to know if there has been any progress with this ~

@vvvm23   Yes, we used this one.

```
# Download the tokenizer weight from OpenAI's DALL-E
TOKENIZER_PATH=/path/to/save/dall_e_tokenizer_weight
mkdir -p $TOKENIZER_PATH
wget -o $TOKENIZER_PATH/encoder.pkl https://cdn.openai.com/dall-e/encoder.pkl
wget -o $TOKENIZER_PATH/decoder.pkl https://cdn.openai.com/dall-e/decoder.pkl
```

**Comment by LiweiPeng** [comment]
@NielsRogge, @addf400 ,
I used the same way in NielsRogge's notebook to use BEiT Model. However, all my MLM accuracy is 0 (i.e. none of the visual tokens it predicts correspond to the ground-truth visual tokens from DALL-E's tokenizer.) 

I used the same DALL-E encoder/decoder as listed above. 

Can you provide update on this issue: Why no token is predicted correctly? Is this expected? Thanks.

**Comment by eliahuhorwitz** [comment]
@LiweiPeng @NielsRogge @addf400 
Have any of you been able to figure this out? 
I am getting similar results, the predictions for the masked areas are wrong and as a result the reconstructed (decoded) image has wrong and random values in the masked areas.

**Comment by LiweiPeng** [comment]
> @LiweiPeng @NielsRogge @addf400 Have any of you been able to figure this out? I am getting similar results, the predictions for the masked areas are wrong and as a result the reconstructed (decoded) image has wrong and random values in the masked areas.

Based on my tests, my 'bad' results were expected for BeiT model. I used linear probing when I did my tests. Linear probing using BeiT base model doesn't work well. The reviewed BeiT paper on  https://openreview.net/pdf?id=p-BhZSz59o4 Table 9 has some detailed results.

**Comment by eliahuhorwitz** [comment]
> > @LiweiPeng @NielsRogge @addf400 Have any of you been able to figure this out? I am getting similar results, the predictions for the masked areas are wrong and as a result the reconstructed (decoded) image has wrong and random values in the masked areas.
> 
> Based on my tests, my 'bad' results were expected for BeiT model. I used linear probing when I did my tests. Linear probing using BeiT base model doesn't work well. The reviewed BeiT paper on https://openreview.net/pdf?id=p-BhZSz59o4 Table 9 has some detailed results.

I'm not sure I understand, the linear probing is related to the final class prediction. The wrong predictions I am seeing are of the tokens, I would expect the reconstruction of correct tokens to work even with the pretrained features as no class specific information is needed for this.

**Comment by LiweiPeng** [comment]
Sorry I am not clear in my previous post. In my case, I am using linear probing for final class prediction. And linear probing result is not good compared to some other models like DINO. For your specific problem, I am not sure.
